snuba/query/allocation_policies/bytes_scanned_rejecting_policy.py

```diff
@@ -0,0 +1,259 @@
+
+
+# we don't limit the amount of bytes subscriptions can scan at this time
+_PASS_THROUGH_REFERRERS = set(
+    [
+        "subscriptions_executor",
+    ]
+)
+
+
+UNREASONABLY_LARGE_NUMBER_OF_BYTES_SCANNED_PER_QUERY = int(1e12)
+_RATE_LIMITER = RedisSlidingWindowRateLimiter(
+    get_redis_client(RedisClientKey.RATE_LIMITER)
+)
+DEFAULT_OVERRIDE_LIMIT = -1
+DEFAULT_BYTES_SCANNED_LIMIT = 10000000
+
+
+class BytesScannedRejectingPolicy(AllocationPolicy):
+    """For every query that comes in, keep track of the amount of bytes scanned for every
+        (project_id|organization_id, referrer)
+    combination in the last 10 minutes (sliding window). If a specific combination scans too
+    many bytes, reject that query
+
+    cross-project queries use the organization_id, single project queries use the project_id
+    """
+
+    WINDOW_SECONDS = 10 * 60
+    WINDOW_GRANULARITY_SECONDS = 60
+
+    def _additional_config_definitions(self) -> list[AllocationPolicyConfig]:
+        # Overrides are prioritized in order of specificity.
+        # If two overrides applicable available to the request, the one with a smaller value takes precedence
+        return [
+            AllocationPolicyConfig(
+                "referrer_all_projects_scan_limit_override",
+                f"Specific referrer scan limit in the last {self.WINDOW_SECONDS/ 60} mins, APPLIES TO ALL PROJECTS",
+                int,
+                DEFAULT_OVERRIDE_LIMIT,
+                param_types={"referrer": str},
+            ),
+            AllocationPolicyConfig(
+                "referrer_all_organizations_scan_limit_override",
+                f"Specific referrer scan limit in the last {self.WINDOW_SECONDS/ 60} mins, APPLIES TO ALL ORGANIZATIONS",
+                int,
+                DEFAULT_OVERRIDE_LIMIT,
+                param_types={"referrer": str},
+            ),
+            AllocationPolicyConfig(
+                "project_referrer_scan_limit",
+                f"DEFAULT: how many bytes can a project scan per referrer in the last {self.WINDOW_SECONDS/ 60} mins before queries start getting rejected",
+                int,
+                DEFAULT_BYTES_SCANNED_LIMIT,
+            ),
+            AllocationPolicyConfig(
+                "organization_referrer_scan_limit",
+                f"DEFAULT: how many bytes can an organization scan per referrer in the last {self.WINDOW_SECONDS/ 60} mins before queries start getting rejected. Cross-project queries are limited by organization_id",
+                int,
+                DEFAULT_BYTES_SCANNED_LIMIT * 2,
+            ),
+        ]
+
+    def _are_tenant_ids_valid(
+        self, tenant_ids: dict[str, str | int]
+    ) -> tuple[bool, str]:
+        if self.is_cross_org_query(tenant_ids):
+            return True, "cross org query"
+        if tenant_ids.get("referrer") is None:
+            return False, "no referrer"
+        return True, ""
+
+    def _get_customer_tenant_key_and_value(
```

Comment:
Why does this return a project id OR an org id? Wouldn't we want to check both levels?

Reply:
To be honest because the ConcurrentLimitAllocationPolicy does the same thing. 

If there's a project_id in the tenant_ids that means one project is being queried so we consume the project quota.
If there's no project_id in the tenant_ids that means many projects are being queried, so we consume the organization quota